core/rawdb/database.go

```diff
@@ -341,46 +388,67 @@ func InspectDatabase(db ethdb.Database) error {
 		}
 	}
 	// Inspect append-only file store then.
-	ancients := []*common.StorageSize{&ancientHeaders, &ancientBodies, &ancientReceipts, &ancientHashes, &ancientTds}
+	ancientSizes := []*common.StorageSize{&ancientHeadersSize, &ancientBodiesSize, &ancientReceiptsSize, &ancientHashesSize, &ancientTdsSize}
 	for i, category := range []string{freezerHeaderTable, freezerBodiesTable, freezerReceiptTable, freezerHashTable, freezerDifficultyTable} {
 		if size, err := db.AncientSize(category); err == nil {
-			*ancients[i] += common.StorageSize(size)
+			*ancientSizes[i] += common.StorageSize(size)
 			total += common.StorageSize(size)
 		}
 	}
+	// Get number of ancient rows inside the freezer
+	ancients := counter(0)
+	if count, err := db.Ancients(); err == nil {
+		ancients = counter(count)
+	}
+	// Count receipts in ancient db
+	ancientReceipts := counter(0)
+	for blockNumber := uint64(0); blockNumber < uint64(ancients); blockNumber++ {
+		data, err := db.Ancient(freezerReceiptTable, blockNumber)
+		if err != nil {
+			log.Error("Error reading ancient receipts from block", "number", blockNumber, "err", err)
+		} else {
+			ancientReceipts += countReceiptsRLP(data)
+		}
+		if blockNumber%1000 == 0 && time.Since(logged) > 8*time.Second {
+			log.Info("Counting ancient database receipts", "number", blockNumber, "percentage", ancients.Percentage(blockNumber), "elapsed", common.PrettyDuration(time.Since(start)))
+			logged = time.Now()
+		}
```

Comment:
Hm... I'm not fond of this.. So, in the 'ideal' case, all ancients are kept in lock-step, so we have identical number of headers, receipts etc. 
In this case, you've found that it's not the case, and when one does a `geth inspect`, it's good that we don't just trust the 'ideal case' but actually verify. However, 

1. if we _do_ verify, I don't see why we should treat receipts as a special snowflake, but do the same of all of them, 
2. There is a better way to check the items instead of reading all the data from disk, namely to expose `t.items` from the `freezerTable`, as maybe `Count()`

Reply:
I think I'm not explained well. We found discrepancies in the tx index for the levelDB store, but ancient store is fine and we didn't find any discrepancy in numbers of headers, bodies, etc. vs receipts.

The PR uses `db.Ancients()` already to count elements from freezer, and the results correspond exactly to the number of blocks saved in the freezer. But - unlikely with levelDB - in freezer receipts are packed in blocks. Because info about levelDB and freezer are in the same report, that count can be misleading to user.

This is why in my opinion _Receipts_ value should be treated differently from the others, because usually we see them as single elements and not as packet of receipts, and because in the key-value store stat we gave just few line above they are counted as separated items.

At the start the following would have been the result based only on `db.Ancients()`

```zsh
+-----------------+--------------------+------------+----------+
|    DATABASE     |      CATEGORY      |    SIZE    |  COUNT   |
+-----------------+--------------------+------------+----------+
| Key-Value store | Receipts           | 39.90 MiB  |    83630 |
...
| Ancient store   | Bodies             | 1.09 GiB   |  3276800 |
| Ancient store   | Receipts           | 596.63 MiB |  3276800 |
| Ancient store   | Difficulties       | 31.21 MiB  |  3276800 |
...
```

But that _Receipts_ count seems misleading to me. Changing the label to _Recepit Blocks_ would have done the trick, but it would have not add much information to how many receipts there are in the ancient, so from there the choice to count receipts with an ad hoc procedure.

Maybe we can opt to use a separate value to preserve that info also, but I don't like very much this method because size column sounds misleading in every case:

```zsh
+-----------------+--------------------+------------+----------+
|    DATABASE     |      CATEGORY      |    SIZE    |  COUNT   |
+-----------------+--------------------+------------+----------+
| Key-Value store | Receipts           | 39.90 MiB  |    83630 |
...
| Ancient store   | Receipt Blocks     | 0.0 B      |  3276800 |
| Ancient store   | Receipts           | 596.63 MiB |  3761928 |
```
```zsh
+-----------------+--------------------+------------+----------+
|    DATABASE     |      CATEGORY      |    SIZE    |  COUNT   |
+-----------------+--------------------+------------+----------+
| Key-Value store | Receipts           | 39.90 MiB  |    83630 |
...
| Ancient store   | Receipt Blocks     | 596.63 MiB |  3276800 |
| Ancient store   | Receipts           | 596.63 MiB |  3761928 |
```

What do you think, does all this make sense for you?

Reply:
I think it makes more sense to change the label, so that for leveldb keep `Receipts`, as in "individual per-transaction receipts", but change `Ancient store | Receipt lists| size | number of items | ` (or alternatively `Receipt Blocks`). 

Quirky sidenote: `3845543-83630-3761928 = -15` == `(number of transactiions indexed) - (number of receipts in leveldb) - (number of receipts in anceintdb) == -15`. I guess there's some data which overlaps and is present both in ancients and leveldb when you ran it.

Reply:
> I think it makes more sense to change the label, so that for leveldb keep `Receipts`, as in "individual per-transaction receipts", but change `Ancient store | Receipt lists| size | number of items | ` (or alternatively `Receipt Blocks`).

Ok, but this way you don't know how many receipts are really stored in ancient db, so you cannot easily do consistency analysis like you did below :) What if we put both info in the report, something like this (changed column title and added counted receipts)?

```zsh
+-----------------+--------------------+------------+----------+
|    DATABASE     |      CATEGORY      |    SIZE    |  ITEMS   |
+-----------------+--------------------+------------+----------+
| Key-Value store | Headers            | 15.00 MiB  |    24295 |
| Key-Value store | Bodies             | 44.52 MiB  |    24295 |
| Key-Value store | Receipts           | 39.90 MiB  |    83630 |
...
| Ancient store   | Receipt lists      | 596.63 MiB |  3276800 |
| Ancient store   | └ counted receipts | --         |  3761928 |
| Ancient store   | Difficulties       | 31.21 MiB  |  3276800 |
...
```

> Quirky sidenote: `3845543-83630-3761928 = -15` == `(number of transactiions indexed) - (number of receipts in leveldb) - (number of receipts in anceintdb) == -15`. I guess there's some data which overlaps and is present both in ancients and leveldb when you ran it.

Interesting. Maybe during chain reorgs the tx index is cleaned up but old receipts are kept in db anyway? I'll try to understand.